sc2: Adding virtual item tracking for weapon/armour upgrades for performance

MatthewMarinets opened the pull request:
## What is this fixing or adding?
Beginning to address the large performance nosedive after PR #5312.

Reported by Mysteryem earlier today in [#ap-core-dev](https://discord.com/channels/731205301247803413/731214280439103580/1413578906912555038).

More can still be done to get the performance back in line, but I'll keep this PR focused on performance boosts from virtual items / overloading `collect()` / `remove()`. (Performance boost from options change now up on PR #5424).

### Weapon/Armour upgrades are now managed by a virtual item
Note there's a slight behavioural change on the logic here: Protoss shield upgrades are no longer covered by logic. As a player I'd say it's fine, shield upgrades are by far the least useful generic upgrades and it was only being checked for in places that were also checking for weapon/armour, so I'd say it's fine and shouldn't lead to any logic holes. The reason shields were left out was because I didn't want to add complexity to hot code to account for the behaviour when the upgrade bundling was set to air/ground -- in that situation, shield upgrades are set to the maximum value of the two upgrade packs.

### Terran defense rating is now regulated by a virtual item
After w/a upgrades were made into virtual functions, the next most expensive rules under the profiler were `terran_competent_comp` and `terran_defense_rating`. `competent_comp` is a far more complex beast, so I aimed for the lower-hanging fruit. I also simplified the logic around it somewhat as it wasn't really matching what I'd expect from a "defense rating". I ran this by the sc2-dev discord chat to make sure it wasn't coming out of left field. The new (and what I expected was old) focus is on specifically items that help you defend even if you F2 your army. This means buildings and units that deploy/siege somehow.

This meant some AA tools and splash tools were removed, like Banshees, splash vikings and valkyries. I looked through all the usages of the function to make sure they still made sense under the new rules and had a reasonable number of permissible combinations of units even under vanilla_items_only.

This refactor and simplification only saved about 2s out of ~220s in my 5-yaml setup under cProfile. There's a few other low-hanging fruit, like terran power level, but I think it's safe to say the payoff is probably too small to promise it as part of this PR.

## How was this tested?
* Ran all unit tests.
* Ran a test generation with 5 template yamls. Generation time went from ~89s to ~68 seconds
* Ran many generations with 5 template yamls under cProfile. Checked time savings were really coming from the functions that were refactored

## If this makes graphical changes, please attach screenshots.
None

Silvris commented on worlds/sc2/pool_filter.py:
```diff
@@ -105,18 +108,36 @@ def copy_item(item: StarcraftItem) -> StarcraftItem:
+
+def after_remove_item(inventory: Counter[str], item: "Item") -> None:
+    for effect in LOGIC_EFFECTS.get(item.code, ()):
+        inventory[effect.name] += 1
```
Shouldn't this be `inventory[effect.name] -= 1`?

Mysteryem commented:
I think now would be a good time to replace the reinventing of the wheel that is `ValidInventory`, and to just use `CollectionState`. That way, any `collect/remove` overrides on the world are supported effectively for free without having to reimplement them for `ValidInventory`, and any future `LogicMixin` would also be supported effectively for free.

I have implemented this on current main in https://github.com/Mysteryem/Archipelago-ahit/commit/2b8c845699bcf789ef92734cf3e55dd7f687d480

One subtle difference is that `CollectionState.__init__()` collects all items in `multiworld.precollected_items`. If that needs to not happen, then `ValidInventory.__init__()` could empty `self.logical_inventory`/`self.prog_items[world.player]` after the `super().__init__()` call.

Ziktofel reviewed (changes requested):
Check the comments, there's also a lead how to handle P shields (it could get a bit slower for just one option used but you still reduce the amount of item types actually queried)

The thing is that having a usable ranged splash unit (even if mobile) makes defensive missions easier. In missions like The Dig and All-in you barely ever F2

Kerrigan levels in other hand could get easily optimized too (but they're used for only a few missions). And I bet they can get slow if there's many 1-level items.

T competent comp will get faster itself by having faster W/A check. But the body of it isn't a simple list or counter check

Generally looks good, I'd add some comments for clarity

MatthewMarinets commented on worlds/sc2/item/virtual_items.py:
```diff
@@ -0,0 +1,164 @@
+    item_table[item_names.PROGRESSIVE_ZERG_GROUND_UPGRADE].code: LogicEffect.ZERG_UPGRADE,
+    item_table[item_names.PROGRESSIVE_ZERG_FLYER_UPGRADE].code: LogicEffect.ZERG_UPGRADE,
+    item_table[item_names.PROGRESSIVE_ZERG_WEAPON_ARMOR_UPGRADE].code: LogicEffect.ZERG_UPGRADE,
+    item_table[item_names.PROGRESSIVE_PROTOSS_GROUND_WEAPON].code: LogicEffect.PROTOSS_UPGRADE,
+    item_table[item_names.PROGRESSIVE_PROTOSS_GROUND_ARMOR].code: LogicEffect.PROTOSS_UPGRADE,
+    item_table[item_names.PROGRESSIVE_PROTOSS_AIR_WEAPON].code: LogicEffect.PROTOSS_UPGRADE,
+    item_table[item_names.PROGRESSIVE_PROTOSS_AIR_ARMOR].code: LogicEffect.PROTOSS_UPGRADE,
+    item_table[item_names.PROGRESSIVE_PROTOSS_WEAPON_UPGRADE].code: LogicEffect.PROTOSS_UPGRADE,
+    item_table[item_names.PROGRESSIVE_PROTOSS_ARMOR_UPGRADE].code: LogicEffect.PROTOSS_UPGRADE,
+    item_table[item_names.PROGRESSIVE_PROTOSS_GROUND_UPGRADE].code: LogicEffect.PROTOSS_UPGRADE,
+    item_table[item_names.PROGRESSIVE_PROTOSS_AIR_UPGRADE].code: LogicEffect.PROTOSS_UPGRADE,
+    item_table[item_names.PROGRESSIVE_PROTOSS_WEAPON_ARMOR_UPGRADE].code: LogicEffect.PROTOSS_UPGRADE,
```
I don't. Seems I missed that, will add it momentarily.

MatthewMarinets replied:
Turns out Quatro is more troublesome than expected, especially for the min counters. I'd have to special-case the protoss upgrades somehow because the algorithm for recalculating a min() would be different:

Normal add: `aggregate_effect = min(component effects)`
Normal remove: `if aggregate_effect > current_effect: aggregate_effect = current_effect`
Protoss add upgrade: `aggregate_effect = min(component_effects) + quatro_value`
Protoss add Quatro: `aggregate_effect += 1`
Protoss remove upgrade: `if aggregate_effect > current_effect + quatro_value: aggregate_effect = current_effect + quatro_value`
Protoss remove quatro: `aggregate_effect -= 1`

I think I'll just add the Quatro check to rules.py

MatthewMarinets commented on worlds/sc2/rules.py:
```diff
@@ -3220,7 +3182,7 @@ def trouble_in_paradise_requirement(self, state: CollectionState) -> bool:
             and self.nova_splash(state)
             and self.terran_beats_protoss_deathball(state)
-            and self.terran_defense_rating(state, True, True) >= 7
+            and self.terran_defense_rating(state, True, True) >= 6
```
This was part of the big audit of all usages of this function, as several options were removed. In practice, this mission does not take a lot of defense rating -- the starting bunker can hold zerg with 1~2 SCVs to repair for the whole mission on brutal. Your army is between you and the protoss for basically the whole mission.

Defense rating used to add a whole bunch of mobile options; with those gone, this check is a lot less relevant.

MatthewMarinets replied:
Yes, and 6 defense rating is still plenty for that. That's tanks + some other static defense, or bunkers with good units to put in them + a turret, or a turret + planetary fortress. I think the only vanilla NCO comp in-logic even at 6 is tanks + bunkers, and even then, NCO has weaker tanks so AP guarantees something better.

I maintain that 6 is sufficient.

MatthewMarinets commented on worlds/sc2/rules.py:
```diff
@@ -1114,7 +1079,7 @@ def kerrigan_levels(self, state: CollectionState, target: int, story_levels_avai
         # Levels from items
         for kerrigan_level_item in kerrigan_levels:
-            level_amount = get_full_item_list()[kerrigan_level_item].number
+            level_amount = item_table[kerrigan_level_item].number
```
I was going by the profiler ratings, and on a 5 default yamls, `kerrigan_levels()` has a cumulative runtime of ~0.15s, less than `morph_brood_lord`, `supreme_requirement()`, `terran_havens_fall_requirement()`, and frankly most named functions in rules.py.

It's an easy optimization for sure, but I'd prefer to keep this PR focused on impactful optimizations first to minimize the chances of breaking things by accident, and we can get the easy ones next dev cycle.

MatthewMarinets commented on worlds/sc2/rules.py:
```diff
@@ -760,32 +726,34 @@ def zerg_defense_rating(self, state: CollectionState, zerg_enemy: bool, air_enem
-        return min(
-            self.weapon_armor_upgrade_count(item_names.PROGRESSIVE_ZERG_MELEE_ATTACK, state),
-            self.weapon_armor_upgrade_count(item_names.PROGRESSIVE_ZERG_GROUND_CARAPACE, state),
+        result = min(
+            state.count(LogicEffect.ZERG_MELEE_ATTACK.name, self.player),
+            state.count(LogicEffect.ZERG_GROUND_ARMOR.name, self.player),
         )
```
I'd advise against it. Both because shield upgrades suck (they're probably the worst generic upgrade), and also because this code also isn't great. I wrote it before reviewing Salz's PR and running some other tests and found `min()` was something like 10x slower than an `if`.

MatthewMarinets replied:
I don't see a problem with shield upgrades being booted out of short worlds. Units and macro upgrades are more valuable anyways.

MatthewMarinets commented:
> The thing is that having a usable ranged splash unit (even if mobile) makes defensive missions easier. In missions like The Dig and All-in you barely ever F2

If that's a concern, it should be on the logic for those missions in particular. Mobile units aren't defense score, it's not what players will expect it to mean when they ask about logic, and it's not what devs will expect when they reach for a pre-made logic function to fill in logic for a new location. The usage I'd expect as a dev/logic writer is "Bunkers are really helpful in this mission, but I don't want to require one particular item; let me use this grab-bag of bunker-like things".

All-In logic already checks things like Valkyries in Air version and Banshee/Battlecruiser on All-In ground. The Dig checks for competent AA and common units. The mobile part is already accounted for, it should not be factored into domain-specific functions. I think this is already handled.

MatthewMarinets commented:
> I think now would be a good time to replace the reinventing of the wheel that is `ValidInventory`, and to just use `CollectionState`. That way, any `collect/remove` overrides on the world are supported effectively for free without having to reimplement them for `ValidInventory`, and any future `LogicMixin` would also be supported effectively for free.
> 
> I have implemented this on current main in [Mysteryem@2b8c845](https://github.com/Mysteryem/Archipelago-ahit/commit/2b8c845699bcf789ef92734cf3e55dd7f687d480)
> 
> One subtle difference is that `CollectionState.__init__()` collects all items in `multiworld.precollected_items`. If that needs to not happen, then `ValidInventory.__init__()` could empty `self.logical_inventory`/`self.prog_items[world.player]` after the `super().__init__()` call.

I agree that the current `ValidInventory` is a bit of a hack (that class has been around since time immemorial and I assume it's just that way because it predates core's current class setup). But I'm uncomfortable refactoring that in this PR.

I'm not very familiar with working with core's `CollectionState` directly, and whenever I have looked into it I've seen indication of undesirable side-effects like the pre-collected dynamic you mention and whatever `prevent_sweep` is doing. Inheriting directly also means we're susceptible to new/changed side-effects if core ever changes. The current setup is a functionality whitelist rather than a blacklist, and I think that's appropriate for this code. As for logic mixins, I think we used to have one for sc2, no one understood it, Berserker commented it was useless and fewer games should use it, and so we removed it. To me, that indicates the trend is towards supporting _fewer_ `LogicMixin`s, not more.

As a programmer, I'm also just very against inheritance in general. Maybe it's the hellish experience of seeing what Java programmers can do to a C codebase to make it more "object oriented". If there's an option that avoids inheritance, I'm taking it (and Python >= 3.10 typing provides this handy-dandy `Protocol` class that lets me get all the perks of interfaces with none of the downsides).

This isn't the only place we use a protocol-like trick with collection state; there's `TestInventory` in test_rules.py. If a refactor of `ValidInventory` finds some good improvements/new patterns that we want, it should be ported there as well. But that's a case of us wanting the whitelist pattern as well; if someone uses a previously-unused CollectionState function in rules.py, we want to manually track that in `TestInventory` to make sure we're adequately checking if everything is progression properly. And that test itself probably wants a refactor (it's currently more of a fuzzer and is quite slow; I've been advocating for more of an AST-parse approach to hopefully speed it up).

Given this PR has already uncovered an uncomfortable number of bugs, I'd rather keep it focused on fixing those and improving performance without throwing in a refactor that may introduce new bugs. I'm saving a link to your commit so we can use it as a jumping-off point when that refactor happens (next dev cycle, Ziktofel has already slated it as a maintenance-focused one).

tl;dr I agree the class needs a refactor, but won't put it in this PR. Next sc2 dev cycle is maintenance-focused, will do the refactor there.

MatthewMarinets commented on worlds/sc2/item/virtual_items.py:
```diff
@@ -0,0 +1,164 @@
+    item_table[item_names.PLANETARY_FORTRESS].code: (LogicEffect.TERRAN_DEFENSE_RATING, 3),
+    item_table[item_names.PERDITION_TURRET].code: (LogicEffect.TERRAN_DEFENSE_RATING|LogicEffect.TVZ_DEFENSE_RATING, 2),
+    item_table[item_names.DEVASTATOR_TURRET].code: (LogicEffect.TERRAN_DEFENSE_RATING, 2),
+    item_table[item_names.LIBERATOR].code: (LogicEffect.TERRAN_DEFENSE_RATING, 3),
```
Set this back to 4 and put an explicit `if` to apply the -2 vs zerg in `terran_competent_comp`.

Snarkie commented:
Tested this locally, the generation time was reduced by about 26% avg on my end. Refer to 
https://docs.google.com/spreadsheets/d/1jijO6xtJgNhAriOd0obWX6b-388gdSzuRHbl8nvFWJM
for more details.

Ziktofel commented:
Re shields: They can be safely ignored if the option is set to bundle_unit_class as the armor check will happen anyways and shield level can't be any worse under these conditions (attach a comment explaining that). In any other mode the pump up/down for virtual items will work

Ziktofel commented:
The problem is with following YAML:

```yam
name: Player{number}
game: Starcraft 2
requires:
  version: 0.6.4 # Version of Archipelago required for this yaml to work as expected.

Starcraft 2:

  progression_balancing: disabled
  accessibility: full
  start_inventory:
    # Start with these items.
    {}

  #######################
  # Difficulty Settings #
  #######################
  game_difficulty: normal
  game_speed: default
  starter_unit: balanced
  required_tactics: standard
  war_council_nerfs: false
  difficulty_curve: standard
  mission_order: mini_campaign
  maximum_campaign_size: 40

  enabled_campaigns: ['Heart of the Swarm']
  enable_race_swap: disabled
  shuffle_no_build: true
  key_mode: progressive_questlines
  shuffle_campaigns: true
  all_in_map: air
  two_start_positions: false
  selected_races:
    # Pick which factions' missions and items can be shuffled into the world.
    ['Zerg', 'Terran', 'Protoss']

  exclude_very_hard_missions: default
  mission_race_balancing: semi_balanced
  ensure_generic_items: 25
  min_number_of_upgrades: 2
  max_number_of_upgrades: -1
  max_upgrade_level: 4
  generic_upgrade_missions: 0
  generic_upgrade_research: auto_in_no_build
  generic_upgrade_research_speedup: false
  generic_upgrade_items: individual_items
  kerrigan_presence: vanilla
  grant_story_levels: disabled
  kerrigan_levels_per_mission_completed: 0
  kerrigan_levels_per_mission_completed_cap: -1
  kerrigan_level_item_sum: 86
  kerrigan_level_item_distribution: size_14
  kerrigan_total_level_cap: -1
  start_primary_abilities: 0
  kerrigan_primal_status: vanilla
  kerrigan_max_active_abilities: 12
  kerrigan_max_passive_abilities: 5
  spear_of_adun_presence: vanilla
  spear_of_adun_present_in_no_build: false
  spear_of_adun_passive_ability_presence: vanilla
  spear_of_adun_passive_present_in_no_build: false
  spear_of_adun_max_active_abilities: 13
  spear_of_adun_max_passive_abilities: 3
  nova_max_weapons: 5
  nova_max_gadgets: 5
  nova_ghost_of_a_chance_variant: wol
  enable_morphling: false
  mercenary_highlanders: false
  victory_cache: 0
  vanilla_locations: enabled
  extra_locations: enabled
  challenge_locations: enabled
  mastery_locations: disabled
  basebust_locations: enabled
  speedrun_locations: disabled
  preventative_locations: enabled
  filler_percentage: 0
  minerals_per_item: 25
  vespene_per_item: 25
  starting_supply_per_item: 2
  maximum_supply_per_item: 1
  maximum_supply_reduction_per_item: 1
  lowest_maximum_supply: 180
  research_cost_reduction_per_item: 2
  filler_items_distribution:
    # Controls the relative probability of each filler item being generated over others.
    # Items that are bound to specific race or option are automatically eliminated.
    # Kerrigan levels generated this way don't go against Kerrigan level item sum
    1 Kerrigan Level: 0
    Additional Maximum Supply: 1
    Additional Starting Minerals: 1
    Additional Starting Supply: 1
    Additional Starting Vespene: 1
    Decreased Maximum Supply: 0
    Increased Building Construction Speed: 1
    Increased Shield Regeneration: 1
    Increased Upgrade Research Speed: 1
    Reduced Upgrade Research Cost: 1

  ###########################
  # Inclusions & Exclusions #
  ###########################
  locked_items:
    # Guarantees that these items will be unlockable, in the amount specified.
    # Specify an amount of 0 to lock all copies of an item.
    {}

  excluded_items:
    # Guarantees that these items will not be unlockable, in the amount specified.
    # Specify an amount of 0 to exclude all copies of an item.
    "Infested Banshee": 0
    "Infested Liberator": 0
    "Mutalisk": 0
    "Corruptor": 0
    "Brood Queen": 0
    "Scourge": 0

  unexcluded_items:
    # Undoes an item exclusion; useful for whitelisting or fine-tuning a category.
    # Specify an amount of 0 to unexclude all copies of an item.
    {}

  vanilla_items_only: false
  exclude_overpowered_items: false
  excluded_missions:
    # Guarantees that these missions will not appear in the campaign
    # Doesn't apply to vanilla mission order.
    # It may be impossible to build a valid campaign if too many missions are excluded.
    []

  #####################
  # Advanced Gameplay #
  #####################
  mission_order_scouting: none
  difficulty_damage_modifier: false
  take_over_ai_allies: false
  enable_void_trade: false
  void_trade_age_limit: 30_minutes
  void_trade_workers: false
  grant_story_tech: no_grant

  ###########################
  # Item & Location Options #
  ###########################
  local_items:
    # Forces these items to be in their native world.
    []

  non_local_items:
    # Forces these items to be outside their native world.
    []

  start_hints:
    # Start with these item's locations prefilled into the ``!hint`` command.
    []

  start_location_hints:
    # Start with these locations and their item prefilled into the ``!hint`` command.
    []

  exclude_locations:
    # Prevent these locations from having an important item.
    []

  priority_locations:
    # Prevent these locations from having an unimportant item.
    []

  item_links:
    # Share part of your item pool with other players.
    []

  plando_items:
    # Generic items plando.
    []
```
Note that these excluded items could be also randomly culled, and you want to require Z air upgrades for competent comp when they get removed because all the Z air units were excluded or culled. For other races this problem is less likely to happen but still can occur. This wasn't a problem before. Also Z competent comp was the rule causing the problem in calculating accessibility for 1000-world generation (challenge objectives and missions requiring this rule failed for one given world)

MatthewMarinets commented:
> Note that these excluded items could be also randomly culled, and you incorrectly want to require Z air upgrades for competent comp

Pushed a fix. I just changed the logic to only require the upgrades necessary for the units currently being matched. I paired it with a rename of `zerg_competent_comp` to `zerg_upgraded_competent_comp` to hopefully make it clearer to future logic writers that it's the function handling upgrade logic. It was already checking for >=2 min upgrade level and was paired with _another_ check of the same thing, so I got to rip out a lot of superfluous checks. Running some performance checks, there was no noticeable drop in the 5 vanilla yaml performance check. The main function being changed is only ~1s of the ~220s generation time and hasn't drifted all that much.

The old flag system really sucks, and I'm glad to be getting rid of it. It essentially lies about what the logic will be until a bunch of stuff is culled, then goes back and changes the logic out after the fact for a later stage. I have no idea what the expected result would be on the culling algorithm, I'd expect a whole bunch of extra stuff would get locked that doesn't have to be, which I'd also expect to lead to more same-y runs.

Ziktofel commented on worlds/sc2/rules.py:
```diff
@@ -949,34 +942,54 @@ def morph_tyrannozor(self, state: CollectionState) -> bool:
+        )
         support_unit = (
-            state.has_any({item_names.SWARM_QUEEN, item_names.HYDRALISK, item_names.INFESTED_BANSHEE}, self.player)
+            state.count(LogicEffect.ZERG_MUTALISK_RATING.name, self.player) >= 14
```
2/2 Mutas weren't considered as competent, there's usually too much stuff in the missions that counter them. This role was always meant for how great the combo of 
```python
item_names.MUTALISK, item_names.MUTALISK_SEVERING_GLAIVE, item_names.MUTALISK_VICIOUS_GLAIVE
```
is, where you do a lot damage with the bounces. Plain 2/2 Mutas + Lings don't work for many missions
Also for attacking supports (like Brood Lord, Hydra, Banshee, Swarm Queen)  there isn't a W/A check. Air shall have 2/2, ranged their 2/2 too

MatthewMarinets replied:
2/2 mutas won't satisfy this check. Only attack upgrades give +1 to the rating, so this is mutas + 4 of (vicious glaive, severing glaive, sundering glaive, +flyer attack ups).

Ziktofel commented:
I don't think that this PR by itself shall redesign logic, just optimize the existing (while keeping the requirements the same).
For redesigning the logic itself it needs. Like the existing logic was made by me and @Magnemania who I treat as logic expert in SC2

Ziktofel commented:
Looks like it only tries to avoid the bug for the Zerg but I think the issue will strike again once I cull or exclude all Terran infantry or vehicles (at least one air unit will probably be locked by logic)

Either way it looks like the rules get very hard to maintain (not that easy to tell what the check against), therefore I've tried an another approach: #5459

MatthewMarinets commented:
> I don't think that this PR by itself shall redesign logic, just optimize the existing (while keeping the requirements the same). For redesigning the logic itself it needs. Like the existing logic was made by me and @Magnemania who I treat as logic expert in SC2. Also we had a round of discussion during the dev cycle for the logic, that resulted in the current logic
> 
> Therefore I'm in favor of the logic optimization but I don't favor smuggling in new logic holes by redefining the location requirements themselves

I wouldn't call patching these "logic holes" -- the logic is currently too restrictive and has always been too restrictive. The logic currently requires +2 ground upgrades even if the logic isn't expecting a player to use ground units, and vice versa. For all the races. And the only patch to this is a special-case around the case where all ground or all air is culled. I consider that harder to maintain over logic that just expresses what the actual rules we're looking for are.

I was also present while we were discussing logic changes. And weapon/armour upgrades were not a major part of them, beyond "this is something that we're going to add this cycle." And when they were added, it was a lazy "minimum level" check everywhere, usually added in too many places so the same redundant check was made multiple times per logic rule. The updated zerg competent comp rule still checks that the main damage-dealer of your comp has the upgrades it needs.

It's possible this extraneous upgrades issue still persists on other factions. And that's inherent to the design -- because the patch is so focused on specifically the case where no units match an upgrade, it will still put in extraneous upgrades if there's only one, non-logically relevant unit in the category. That's not going to be avoidable unless you update the rules, it's inherent to the design.
